### PR TITLE
feat:  Add improved platform support

### LINKED ISSUE #143: Add optional log file path to eclcc preferences
Currently, the eclcc.log file is created in the current working directory.  That can clutter up the code space.  A preference for explicitly setting a log directory, or the log file's full path, would be great.

### LINKED ISSUE #184: Feature request: Pin selected launch.json entry to an ECL file
Pasting from an email:

"I'm currently working on a project with several BWRs, banging up against Dataland.  Some of the jobs should target a thor cluster, others an hthor cluster.  Is it feasible to have an option in VS Code where the BWR remembers which launch.json config was last used and switch to it when the file comes forward?"

### LINKED ISSUE #187: Ability to run multiple workunits at the same time

### LINKED ISSUE #188: Able to see the running/blocked WUs on clusters in the sidebar.
it helps to gauge the load before submitting a new WU.

### LINKED ISSUE #190: The ease to run a BWR in different targets especially when working on the THOR world.

### LINKED ISSUE #191: List of all previously executed workunits.

### PR DESCRIPTION
**Pull request by GordonSmith**
Added Workunit Tree
Added Submit / Compile buttons
Added pinned launch configuration
Added eclcc log file override

Fixes #191
Fixes #188
Fixes #187
Fixes #190
Fixes #143
Fixes #184

Signed-off-by: Gordon Smith <GordonJSmith@gmail.com>

### CODE REVIEW

**Comment by GordonSmith** [comment]
@jbrundage please review

**Comment by GordonSmith** [comment]
@jbrundage please review

File: README.md [hunk 1]
```diff
@@ -63,18 +93,18 @@ The following Visual Studio Code settings are available for the ECL extension.
...
 
-  // Add '-legacy' argument to eclcc.
-  "ecl.legacyMode": false
+  // Write eclcc logfile to specified file.
+  "ecl.eclccLogfile": ""
 
-  // Open Workunits in external browser.
-  "ecl.WUOpenExternal": true
+  // Write eclcc logfile to specified file.
+  "ecl.eclccLogfile": ""
```
Comment (GordonSmith) [inline README.md:100]: Dup

File: README.md [hunk 2]
```diff
@@ -22,25 +37,40 @@ The following ECL specific commands are available.  Note:  These commands will *
...
 #### Within the ECL Code Editor:
 
-* Syntax Check **[F7]** - _Save + check syntax of current file._
-* Language Reference Lookup **[shift + F1]** - _For the currently selected text, search the online ECL language reference._
+| Command                   | Shortcut | Description                                                      |
+|---------------------------|:--------:|------------------------------------------------------------------|
+|![Submit](resources/light-png/play.png) Submit | F5 | Submit ECL |
+|![Compile](resources/light-png/file-binary.png) Compile | | Compile ECL |
+| Syntax Check | F7 | Save and check syntax of current file |
+| Language Reference Lookup | shift + F1 | Toggle between "My" and "All" Workunits |
```
Comment (GordonSmith) [inline README.md:54]: Wrong description

File: README.md [hunk 3]
```diff
@@ -22,25 +37,40 @@ The following ECL specific commands are available.  Note:  These commands will *
...
-#### Within the ECL Activity Pane:
-_Right click on item_
+#### Within the Workunit Tree:
 
-* Workunit Details - _Opens ECL Watch Workunit Details for the selected Workunit._
+| Command                   | Shortcut | Description                                                      |
+|---------------------------|:--------:|------------------------------------------------------------------|
+|![My Workunits](resources/light-png/person.png) My workunits | | Toggle between "My" and "All" Workunits |
+|![All Workunits](resources/light-png/organization.png) All workunits | | Toggle between "My" and "All" Workunits |
+|![Refresh](resources/light-png/refresh.png) ECL Watch | | Refresh Tree |
```
Comment (GordonSmith) [inline README.md:62]: Wrong command string

File: src/ecl/check.ts [hunk 4]
```diff
@@ -74,7 +74,7 @@ function checkUri(uri: vscode.Uri, eclConfig: vscode.WorkspaceConfiguration): Pr
     if (uri) {
         const wsf = vscode.workspace.getWorkspaceFolder(uri);
         if (wsf) {
-            vscode.window.setStatusBarMessage(`Syntax Check:  ${path.relative(wsf.uri.fsPath, uri.fsPath)}`);
+            //             vscode.window.setStatusBarMessage(`Syntax Check:  ${path.relative(wsf.uri.fsPath, uri.fsPath)}`);
```
Comment (GordonSmith) [inline src/ecl/check.ts:77]: Commented code.

File: src/ecl/eclWatchDocument.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import * as vscode from "vscode";
+
+export class ECLWatchEditorProvider implements vscode.CustomReadonlyEditorProvider {
+
+    public static register(context: vscode.ExtensionContext): vscode.Disposable {
+
+        // vscode.workspace.registerTextDocumentContentProvider("cowsay", new MyProvider());
```
Comment (GordonSmith) [inline src/ecl/eclWatchDocument.ts:7]: cowsay!

File: src/ecl/eclWatchDocument.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    openCustomDocument(uri: vscode.Uri, openContext: vscode.CustomDocumentOpenContext, token: vscode.CancellationToken): vscode.CustomDocument | Thenable<vscode.CustomDocument> {
+        throw new Error("XXXMethod not implemented.");
+    }
+
+    resolveCustomEditor(document: vscode.CustomDocument, webviewPanel: vscode.WebviewPanel, token: vscode.CancellationToken): void | Thenable<void> {
+        throw new Error("YYYMethod not implemented.");
+    }
+}
```
Comment (GordonSmith) [inline src/ecl/eclWatchDocument.ts:30]: Remove entire file

File: src/eclwatch.tsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
+import * as React from "react";
+import * as ReactDOM from "react-dom";
+import { WUDetails } from "./eclwatch/WUDetails";
+import { ThemeProvider } from "./eclwatch/themeGenerator";
+
+const bodyStyles = window.getComputedStyle(document.body);
+
+const backColor = bodyStyles.getPropertyValue("--vscode-editor-background");
+const foreColour = bodyStyles.getPropertyValue("--vscode-input-foreground");//Palette.textColor(backColor);
```
Comment (GordonSmith) [inline src/eclwatch.tsx:9]: commented code

File: src/eclwatch.tsx [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    switch (message.command) {
+        case "navigate":
+            render(message.data.launchRequestArgs.protocol, message.data.launchRequestArgs.serverAddress, message.data.launchRequestArgs.port, message.data.launchRequestArgs.user, message.data.launchRequestArgs.password, message.data.wuid, message.data.result);
+            break;
+    }
+});
+
+// const oldState = vscode.getState() || { url: "" };
+// const iframe: HTMLIFrameElement = document.getElementById("myFrame") as any;
+// iframe.src = oldState.url;
```
Comment (GordonSmith) [inline src/eclwatch.tsx:47]: Commented code

File: src/eclwatch.tsx [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        sequence={result}
+        width={clientRect.width}
+        height={clientRect.height}
+    />, placeholder);
+
+}
+
+//  Local debugging without VS Code
+if (document.location.protocol === "file:") {
+    //render("http", "10.173.14.207", "8010", "gosmith", "28Ee1$BHoJvNOTk$KL6F", "W20200910-090123");
```
Comment (GordonSmith) [inline src/eclwatch.tsx:83]: Commented code

File: src/eclwatch/WUResultFluent.tsx [hunk 10]
```diff
@@ -0,0 +1,465 @@
...
+    "labore et dolore magna aliqua ut enim ad minim veniam quis nostrud exercitation ullamco laboris nisi ut " +
+    "aliquip ex ea commodo consequat duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore " +
+    "eu fugiat nulla pariatur excepteur sint occaecat cupidatat non proident sunt in culpa qui officia deserunt "
+).split(" ");
+let loremIndex = 0;
+function _lorem(wordCount: number): string {
+    const startIndex = loremIndex + wordCount > LOREM_IPSUM.length ? 0 : loremIndex;
+    loremIndex = startIndex + wordCount;
+    return LOREM_IPSUM.slice(startIndex, loremIndex).join(" ");
+}
```
Comment (GordonSmith) [inline src/eclwatch/WUResultFluent.tsx:465]: Remove file

File: src/eclwatch/WUResultStore.ts [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+    }
+
+    fetchRange(options): Promise<any[]> {
+        const retVal = new Deferred();
+        this._request(options.start, options.end).then(response => retVal.resolve(response));
+        return new QueryResults(retVal.then(response => response.data), {
+            totalLength: retVal.then(response => response.totalLength)
+        });
+    }
+}
```
Comment (GordonSmith) [inline src/eclwatch/WUResultStore.ts:193]: Check if this custom version is needed.

File: test.html [hunk 12]
```diff
@@ -1,104 +1,16 @@
...
-        });
-    </script>
+<body>
+    <iframe id="myFrame" src="http://10.173.14.207:8010/esp/files/stub.htm"
+        style="position:fixed; top:0; left:0; bottom:0; right:0; width:100%; height:100%; border:none; margin:0; padding:0; overflow:hidden; z-index:999999;">
+        Your browser doesn't support iframes
+    </iframe>
 </body>
 
 </html>
```
Comment (GordonSmith) [inline test.html:16]: Remove

File: webview-test.html [hunk 13]
```diff
@@ -0,0 +1,2083 @@
...
+</head>
+
+<body class="vscode-dark" data-vscode-theme-kind="vscode-dark" data-vscode-theme-name="Dark+ (default dark)">
+    <div id="placeholder">
+    </div>
+    <script nonce="zes9uXypVmEN3lSoLeogIeTyRgDkjQUE" src="./dist/eclwatch.js"></script>
+
+</body>
+
+</html>
```
Comment (GordonSmith) [inline webview-test.html:2083]: Remove

File: webview-styles.txt [hunk 14]
```diff
@@ -0,0 +1,389 @@
...
+--vscode-terminal-ansiBrightMagenta:#d670d6;
+--vscode-terminal-ansiBrightCyan:#29b8db;
+--vscode-terminal-ansiBrightWhite:#e5e5e5;
+--vscode-gitDecoration-addedResourceForeground:#81b88b;
+--vscode-gitDecoration-modifiedResourceForeground:#e2c08d;
+--vscode-gitDecoration-deletedResourceForeground:#c74e39;
+--vscode-gitDecoration-untrackedResourceForeground:#73c991;
+--vscode-gitDecoration-ignoredResourceForeground:#8c8c8c;
+--vscode-gitDecoration-conflictingResourceForeground:#6c6cc4;
+--vscode-gitDecoration-submoduleResourceForeground:#8db9e2;
```
Comment (GordonSmith) [inline webview-styles.txt:389]: Remove

**Review by jbrundage — APPROVED** [review]

**Pull request merged**